PKCS11: Support for Object related API's - Find Objects, GetObjectSize, GetAttributeValue

ruchi393 opened the pull request:
<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

ruchi393 commented:
@vesajaaskelainen , Updated the PR to add commit for C_GetObjectSize() and C_GetAttributeValue(). Please check.

vesajaaskelainen commented:
@ruchi393 I have done some checking and further refined my serialization fixes. Now they seems to be in quite good shape. Should pass now additional tests and usable with the tooling that can be tested with the current state.

Pushed changes to wip branches.

OP-TEE OS:
https://github.com/vesajaaskelainen/optee_os/commits/wip-ec-support

Mainly commits:
- https://github.com/vesajaaskelainen/optee_os/commit/2f952f11d26f59bbd1fc0a685312360179003fc8 pkcs11: fix get attribute data alignment problem
- https://github.com/vesajaaskelainen/optee_os/commit/97b44f0af2a18ff6fcb3dfdec01d11ed4fc2df26 pkcs11: get_attribute: fix returning values into larger buffer
- https://github.com/vesajaaskelainen/optee_os/commit/c015d4e00958c9dbf61d10101696103592413873 pkcs11: Fix serialization handling for non-indirect attributes

OP-TEE client:
https://github.com/vesajaaskelainen/optee_client/commits/wip-ec-support

Mainly commits:
- https://github.com/vesajaaskelainen/optee_client/commit/3ad6ce99eecf4dadf00484c37c62b39ba8a0774e libckteec: deserialize_ck_attributes: Fix handling of attribute serializer
- https://github.com/vesajaaskelainen/optee_client/commit/6201010d6c23b270b783597439914b200eb6239e libckteec: deserialize_ck_attribute: Fix return of attribute data size
- https://github.com/vesajaaskelainen/optee_client/commit/89d274c92af1243e6f894f4519df76fd13ddbe98 libckteec: deserialize_ck_attributes: Fix data alignment
  - If your Linux kernel is configured with data alignment fixups you don't see much effect for this commit

Next question is how shall we handle these fixes -- I can separate them for different PR or you can cherry-pick them.

There was no yet tests for get attribute but once you have skeleton there I can help with some cases that handle those. Basically I have used different tools to test it out and then made capture with pkcs11-spy and then started figuring out what is the problem.

ruchi393 commented:
Thanks @vesajaaskelainen . I will have a look at these and get back to you.

vesajaaskelainen commented on ta/pkcs11/src/object.h:
```diff
@@ -56,4 +56,21 @@ enum pkcs11_rc entry_create_object(struct pkcs11_client *client,
+enum pkcs11_rc entry_get_object_size(struct pkcs11_client *client,
+				     uint32_t ptypes, TEE_Param *params);
+
+void release_session_find_obj_context(struct pkcs11_session *session);
```
add empty line between `release_session_find_obj_context()` and `#endif`.

ruchi393 replied:
> @ruchi393 please note that I have updated wip branch with new commits and updated the comment above.

Thanks @vesajaaskelainen . I was thinking I will work on the comments in this PR first. Once this PR is merged,  you can raise another PR with fixes in the serialization logic on your wip branch. Does that sound reasonable ?

vesajaaskelainen replied:
I am OK with that.

vesajaaskelainen commented:
I assume you will be running `checkpatch` and fixing findings before next update. And I try not to comment on those.

vesajaaskelainen commented on ta/pkcs11/src/attributes.c:
```diff
@@ -207,6 +207,50 @@ bool get_bool(struct obj_attrs *head, uint32_t attribute)
+
+		TEE_MemMove(&pkcs11_ref, ref_attr, sizeof(pkcs11_ref));
+
+		shift = pkcs11_attr2boolprop_shift(pkcs11_ref.id);
```
This boolprop stuff is in @etienne-lms's branch as conditional code but I am not yet convinced that it is actually a good thing to have. It causes problems for serializers. I understand that it makes things slightly more compact but also changes storage and REE interfaces where as storage is more challenging to change.

But the code what you want to do here is actually _is_attribute_bool()_ which could internally use whatever method is in use so shall we have helper for that?

ruchi393 replied:
Makes sense. I will introduce a helper function

etienne-lms replied:
Agree. Thank @vesajaaskelainen.

vesajaaskelainen commented on ta/pkcs11/src/object.c:
```diff
@@ -409,3 +409,591 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+		     read_bytes, res);
+		goto out;
+	}
+	if (read_bytes != info.dataSize) {
```
empty line between if clauses?

ruchi393 replied:
Done

vesajaaskelainen commented on ta/pkcs11/src/object.c:
```diff
@@ -409,3 +409,591 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+		goto out;
+	}
+
+	rc = sanitize_client_object(&req_attrs, template,
```
Should we first sanitize input and after that allocate find_ctx? Eg. switch order of `TEE_Malloc()` and `sanitize_client_object()`.

ruchi393 replied:
Yes, I will do that

etienne-lms replied:
smart, thanks.

ruchi393 commented:
> I assume you will be running `checkpatch` and fixing findings before next update. And I try not to comment on those.

Yes I will be fixing those. I had done that on earlier patch-set but looks like few issues escaped in yesterday's patches. Sorry about that.

vesajaaskelainen commented on ta/pkcs11/src/object.c:
```diff
@@ -409,3 +409,591 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+	ctx = session->find_ctx;
+
+	/*
+	 * TODO: should we check again if these handles are valid?
```
How would you do that?

It is internal structure in session and I suppose it should be good enough if it is just non-NULL. Assuming it is allocated with zero fill.

etienne-lms replied:
I expect it is zero-inited at allocation. I think we can remove this inline comment.

ruchi393 replied:
Agree, I will remove it.

vesajaaskelainen commented on ta/pkcs11/src/object.c:
```diff
@@ -409,3 +409,591 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+	struct serialargs ctrlargs = { };
+	struct pkcs11_session *session = NULL;
+	struct pkcs11_find_objects *ctx = NULL;
+	char *out_handles = NULL;
```
`uint8_t *` to be more specific that it is raw serialized data?

etienne-lms replied:
`char *` and `uint8_t *` are both fine to me.

ruchi393 replied:
Changing it to uint8_t* for clarity

vesajaaskelainen commented on ta/pkcs11/src/object.c:
```diff
@@ -409,3 +409,591 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+			continue;
+
+		/* Newly published handles: store in session list */
+		obj = handle_lookup(&session->object_handle_db,
```
why do we need to do this?

etienne-lms replied:
Hmmm, you're right, not needed. Thanks!

vesajaaskelainen replied:
Then I believe there is no need for `temp_start`? And above if comes redundant.

etienne-lms replied:
`temp_start` is needed by `release_find_obj_context()`. It prevents having entries in the session object database that client does not access. When freeing find context, the handlers already returned shall not be released, only those not fetched by client.
An alternate way could be to keep these objects in the session db until the session is closed, and get rid of `temp_start` as suggested.

In the mean time, here, could be change to:
```c
	for (count = 0, idx = ctx->next; idx < ctx->count, count < out_count;
	     idx++, count++)
		TEE_MemMove(out_handles + count * sizeof(uint32_t),
			    ctx->handles + idx, sizeof(uint32_t));

	ctx->next = idx;
```

etienne-lms replied:
> Unrelated thought -> handle_put() seems to be quiet mis-leading. Somehow the name doesn't suggest we are trying to destroy a handle. Should we change it to handle_destroy()

handle.c exports `handle_get()` and `handle_put()` to create destroy handles. I think __get/put__ are often used  for such purpose.

> Look at the Testcase: (...) So this testcase assumption that same handle will be available across 2 sessions seems wrong to me. Do you agree ?

Yes, I agree, the test should not succeed.

vesajaaskelainen replied:
I have fixes for session usage in my tree's. I will give a PR for those soon and also fix some other session related things that are in master now.

ruchi393 replied:
As I was going through the logic again, there is an issue with the way use of temp_start. In release_find_obj_context(), using temp_start would even erase the handles of the token objects from the session->object_handle_db which were even created by this session making that handle as invalid for that session which is not the right case/
eg.
C_OpenSession()
Create token objects -> save handles
C_FindInit() -> find all token objects.
C_FindFinal()
[In above operation we have skipped the C_Find()]
or a scenario where you try to get all token objects but retrieve only 1 of them using C_Find() and call C_FindFinal().
If we try and use the object handles saved in step 2 above, these would be considered as invalid handles which is not correct.

One way of dealing with this scenario would be to remove the usage of temp_start all together as @etienne-lms suggested.  The only issue with this would be if a token object which was not created by this session was added to database in C_FindInit() but not fetched in C_Find(). We would have an extra object handle with session in this case which application is not aware of. This however would not lead to any error scenario.

So, I am proposing that we remove temp_start all together. @vesajaaskelainen , @etienne-lms , let me know your thoughts.

etienne-lms replied:
I don't catch the issue. In the scenario
> C_OpenSession()
Create token objects -> save handles
C_FindInit() -> find all token objects. - (
C_FindFinal()

the created object handles were already published in the session, so C_FindInit() lists them before `temp_start`.
If there were other token object not already opened by this session, each gets a handle value (`handle_get()`). C_FindFinal() only destroy handles (`handle_put()`) never output to client: from `find_ctx->next` position, at least starting from `temp_start` position.
So `handle_put()` is never called for a handle previously output to the client.

ruchi393 replied:
The logic in C_FindInit() iterates through the session object list first. This would not have any token objects created by the session. Then it iterates over the token object list , checks if handle of token object found exists in the object_handle_db else adds it there. temp_start if set to the index of first token object. Thus even published token objects end up after token_start.

I have also added a test case for same in xtest, which fails with current implementation.

etienne-lms replied:
No, published token objects were already in `session->object_handle_db` since published, so found at 1st pass.
The pass on token object skips those already in `session->object_handle_db` (case they already have a handle).

Thanks for the added test.

etienne-lms replied:
Despite my arguments above, i think it's fine with your simplification, removing the handle release optimization. I would not expect a client keeps a unique pkcs11 session to seek into the many objects and to perform target crypto operations.

ruchi393 replied:
What you are saying is true. However, temp_start indicates the index in find_ctx and is set to find_ctx->count which indicates  only the session objects. The pass on token objects skips adding the token objects already in object_handle_db. However they are all added in the find_ctx->handle.

This the reason why the subcase SubTest 7 I moved in the recent test failed when it tried to find a token object after doing a find operation.

vesajaaskelainen commented on ta/pkcs11/src/object.c:
```diff
@@ -409,3 +409,591 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+	if (out->memref.size != sizeof(uint32_t))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	assert(obj->attributes);
```
I don't think this has any value in here. We have already accessed `obj->attributes` before this.

etienne-lms replied:
ok, we can remove.

ruchi393 replied:
Removed

vesajaaskelainen commented:
@ruchi393 please note that I have updated wip branch with new commits and updated the comment above.

etienne-lms commented on ta/pkcs11/src/attributes.h:
```diff
@@ -179,6 +179,15 @@ static inline enum pkcs11_rc get_u32_attribute(struct obj_attrs *head,
 }
 
+/*
+ * Return true all attributes from the reference are found and match value
```
"Return true **if** all attributes from the reference ..."

ruchi393 replied:
Changed

etienne-lms reviewed (commented):
I looked back in the history of this implementation of this commands and @EDVTAZ is also to credit for these.
I suggest to add to each commit log:
```
Co-developed-by: Gabor Szekely <szvgabor@gmail.com>
Signed-off-by: Gabor Szekely <szvgabor@gmail.com>

ruchi393 commented:
> I looked back in the history of this implementation of this commands and @EDVTAZ is also to credit for these.
> I suggest to add to each commit log:
> 
> ```
> Co-developed-by: Gabor Szekely <szvgabor@gmail.com>
> Signed-off-by: Gabor Szekely <szvgabor@gmail.com>
> ```

Thanks @etienne-lms for mentioning this. I will add the credits.

ruchi393 commented:
Addressed review comments and checkpatch issues till date.

Things remaining:
1.  Following tag to be applied
Co-developed-by: Gabor Szekely <szvgabor@gmail.com>
Signed-off-by: Gabor Szekely <szvgabor@gmail.com>
2. Need to decide on the ID's in pkcs11_ta.h. Should we leave some ID's for GENERATE_KEYPAIR etc or add them in sequence.

etienne-lms commented on ta/pkcs11/src/pkcs11_helpers.h:
```diff
@@ -55,6 +55,9 @@ bool key_type_is_asymm_key(uint32_t key_type_id);
 /* Boolprop flag shift position if @attribute_id is boolean, else -1 */
 int pkcs11_attr2boolprop_shift(uint32_t attribute_id);
 
+/* Return true if @attribute_id is boolean, else false */ 
```
trailing space at end of line.

ruchi393 replied:
Fixed

etienne-lms commented on ta/pkcs11/src/pkcs11_attributes.c:
```diff
@@ -1252,3 +1252,42 @@ check_parent_attrs_against_processing(enum pkcs11_mechanism_id proc_id,
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	enum pkcs11_class_id key_class = get_class(obj->attributes);
+
+	if (key_class != PKCS11_CKO_SECRET_KEY ||
```
bug here :(     should be `&&`, not `||`.

reported by clang thru shippable :)
```
 *  _make COMPILER=clang
ta/pkcs11/src/pkcs11_attributes.c:1264:41: error: overlapping comparisons always evaluate to true [-Werror,-Wtautological-overlap-compare]
        if (key_class != PKCS11_CKO_SECRET_KEY ||
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
1 error generated.

ruchi393 replied:
Good that tooling caught it. I have fixed the issue.

etienne-lms replied:
thanks @jforissier :)

etienne-lms reviewed (changes requested):
few remaining issues reported by CI.

ruchi393 commented:
@etienne-lms , @vesajaaskelainen , I have fixed all comments, shippable and checkpatch errors till now in all the 3 repos -> here, optee_client and optee_test.  Please review.

etienne-lms commented on ta/pkcs11/src/attributes.c:
```diff
@@ -207,6 +207,48 @@ bool get_bool(struct obj_attrs *head, uint32_t attribute)
+	}
+
+	for (count = 0; count < ref->attrs_count; count++) {
+		struct pkcs11_attribute_head pkcs11_ref;
```
` = { };`

ruchi393 replied:
Fixed

etienne-lms commented on ta/pkcs11/src/object.c:
```diff
@@ -409,3 +409,571 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+
+out:
+	TEE_Free(template);
+	template = NULL;
```
can remove this line.

ruchi393 replied:
fixed

etienne-lms commented on ta/pkcs11/src/pkcs11_helpers.c:
```diff
@@ -464,6 +469,17 @@ int pkcs11_attr2boolprop_shift(uint32_t attr)
 
+bool pkcs11_is_attr_bool(uint32_t attr)
+{
+	int shift = 0;
```
suggestion:
```c
bool pkcs11_is_attr_bool(uint32_t attr)
{
	return pkcs11_attr2boolprop_shift(attr) >= 0;
}

ruchi393 replied:
Changed

etienne-lms commented on ta/pkcs11/src/pkcs11_token.h:
```diff
@@ -142,6 +142,23 @@ struct active_processing {
+ * @count - number of matching handle found
+ * @handles - array of handle of matching objects (published handles)
+ * @next - index of the next object handle to return to FindObject
+ * @temp_start - index of the trailing not yet published handles
```
maybe rename it, maybe s/temp_start/no_hdl_start/.

ruchi393 replied:
Removed temp_start all together

etienne-lms reviewed (commented):
some minor findings.
If @vesajaaskelainen agree, I think you can squash the fixup commits.

ruchi393 commented:
@vesajaaskelainen , @etienne-lms thanks a lot for the review. Should I squash the commits now ?

vesajaaskelainen commented on ta/pkcs11/src/attributes.c:
```diff
@@ -207,6 +207,48 @@ bool get_bool(struct obj_attrs *head, uint32_t attribute)
+
+		if (pkcs11_is_attr_bool(pkcs11_ref.id)) {
+			bool bb_ref = get_bool(ref, pkcs11_ref.id);
+			bool bb_candidate = get_bool(candidate, pkcs11_ref.id);
```
When looking at some session handling things I noticed that when using get_bool() there is possible problem in here.

If candidate does not have ref's attribute at all it gets value "false" and if ref's lookup value was "false" it matches.

Now is this a real problem or not I do not yet know. In theory we have proper default values if value is missing when creating an object.

But if one searches for specific entries and that particular attribute is from different object class then this might be problem.

It might be safest to do something like:
```
rc = get_attribute_ptr(candidate, pkcs11_ref.id,
			&found, &size);
if (rc || size != pkcs11_ref.size)
  return false;

if (pkcs11_is_attr_bool(pkcs11_ref.id))
...
else
...
```

I would porbably also rename "found" to be more like "(canditate_)value" or so.

ruchi393 replied:
Good catch @vesajaaskelainen . To clarify if I understand it correctly,  I am giving an example below 
eg 
CK_ATTRIBUTE search_template[] = {
                { CKA_TOKEN, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
                { CKA_ENCRYPT,  &(CK_BBOOL){CK_FALSE}, sizeof(CK_BBOOL) }, \
        };

Trying to search with above template on a data object should return no objects ideally as the template has the property CKA_ENCRYPT in it which is not a valid attribute for data object. Adding this check would return no objects for the above template.

If above understanding is correct, I would agree it is safe to add the check you mentioned above.

etienne-lms replied:
As you said @vesajaaskelainen, the purpose here was to allow qiuck access to identified boolean attributes called boolprops, which covered the generic object standard boolean CKA_xxx. If we assume `get_attribute_ptr()` returns a valid byte pointer to the boolean value, there is no purpose calling `get_bool()`.

All in one, I would agree to simplify to:
```c
for (count = 0; count < ref->attrs_count; count++) {
	struct pkcs11_attribute_head pkcs11_ref = { };
	void *found = NULL;
	uint32_t size = 0;

	TEE_MemMove(&pkcs11_ref, ref_attr, sizeof(pkcs11_ref));

	rc = get_attribute_ptr(candidate, pkcs11_ref.id, &found, &size);
	if (rc || !found || size != pkcs11_ref.size ||
	    TEE_MemCompare(ref_attr + sizeof(pkcs11_ref), found, size))
		return false;

	ref_attr += sizeof(pkcs11_ref) + pkcs11_ref.size;
}

return true;
```

vesajaaskelainen commented on ta/pkcs11/src/attributes.c:
```diff
@@ -207,6 +207,48 @@ bool get_bool(struct obj_attrs *head, uint32_t attribute)
+					       &found, &size);
+
+			if (rc || !found || size != pkcs11_ref.size ||
+			    TEE_MemCompare(ref_attr + sizeof(pkcs11_ref),
```
This makes an assumption about internal storage in memory.

Should we just get data pointer with get_attribute_ptr(ref, ...) and then compare actual attribute value? That should work in all cases should it be refactored at some point of time.

vesajaaskelainen replied:
I believe this is OK -- we anyway use size to advance forward. No change needed.

vesajaaskelainen commented:
> @vesajaaskelainen , @etienne-lms thanks a lot for the review. Should I squash the commits now ?

Check the get_bool() comment with or without that feel free to squash. I will anyway check them as a whole next time.

etienne-lms commented on ta/pkcs11/src/pkcs11_helpers.h:
```diff
@@ -55,6 +55,9 @@ bool key_type_is_asymm_key(uint32_t key_type_id);
 int pkcs11_attr2boolprop_shift(uint32_t attribute_id);
 
+/* Return true if @attribute_id is boolean, else false */
+bool pkcs11_is_attr_bool(uint32_t attribute_id);
```
What about this?
```c
/* Return true if @attribute_id is boolean, else false */
static inline bool pkcs11_is_attr_bool(uint32_t attr)
{
	return pkcs11_attr2boolprop_shift(attr) >= 0;
}
```

ruchi393 replied:
Thanks. Removed the duplicate function pkcs11_is_attr_bool()

etienne-lms commented on ta/pkcs11/src/pkcs11_token.h:
```diff
@@ -142,6 +142,21 @@ struct active_processing {
+ *
+ * @attributes - matching attributes list searched (null if no search)
+ * @count - number of matching handle found
+ * @handles - array of handle of matching objects (published handles)
```
can remove `(published handles)`.

ruchi393 replied:
Changed

etienne-lms replied:
fixed in wrong commit

etienne-lms commented on ta/pkcs11/src/pkcs11_token.h:
```diff
@@ -142,6 +142,21 @@ struct active_processing {
+ * @attributes - matching attributes list searched (null if no search)
+ * @count - number of matching handle found
+ * @handles - array of handle of matching objects (published handles)
+ * @next - index of the next object handle to return to FindObject
```
either `C_FindObjects` or `PKCS11_CMD_FIND_OBJECTS`.

ruchi393 replied:
Changed

ruchi393 commented:
Comments addressed

ruchi393 commented:
@vesajaaskelainen can you help review the squashed commits here as well as in optee_client. I have taken care of all comments till date. It am hoping to get this PR merged for the coming release.

vesajaaskelainen commented:
> @vesajaaskelainen can you help review the squashed commits here as well as in optee_client. I have taken care of all comments till date. It am hoping to get this PR merged for the coming release.

Sure. I would like to do my test battery for that. But that goes to later today.

etienne-lms commented:
Hi, I've found some issues in the current implementation.
@ruchi393, could look at the 3 commits below and cherry-pick then into this P-R if you agree on content.
https://github.com/etienne-lms/optee_os/commit/642c7e4b1ccb854c1be9cb38aa576f23956eccb8
https://github.com/etienne-lms/optee_os/commit/4509c6554eabfce81cf4259555baafb944aef7bc
https://github.com/etienne-lms/optee_os/commit/b5d5beb3018412c32c30064b052f7699502bee2c

ruchi393 commented:
> Hi, I've found some issues in the current implementation.
> @ruchi393, could you look at the 3 commits below and cherry-pick then into this P-R if you agree on content?
> [etienne-lms@642c7e4](https://github.com/etienne-lms/optee_os/commit/642c7e4b1ccb854c1be9cb38aa576f23956eccb8)
> [etienne-lms@4509c65](https://github.com/etienne-lms/optee_os/commit/4509c6554eabfce81cf4259555baafb944aef7bc)
> [etienne-lms@b5d5beb](https://github.com/etienne-lms/optee_os/commit/b5d5beb3018412c32c30064b052f7699502bee2c)

@etienne-lms , thanks for the patches. These look good to me. I have cherry-picked them in the PR. For checking etienne-lms@642c7e4, I have also added a testcase which now passes after this fix.
https://github.com/OP-TEE/optee_test/pull/474

vesajaaskelainen commented on ta/pkcs11/src/attributes.c:
```diff
@@ -207,6 +207,37 @@ bool get_bool(struct obj_attrs *head, uint32_t attribute)
+
+	for (count = 0; count < ref->attrs_count; count++) {
+		struct pkcs11_attribute_head pkcs11_ref = { };
+		void *found = NULL;
```
Could we rename `found` as `value` or so?

With name `found` I kinda first think it as boolean.

etienne-lms replied:
agree

ruchi393 replied:
Changed "found" to "value"

vesajaaskelainen commented on ta/pkcs11/src/object.c:
```diff
@@ -427,3 +427,552 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+out:
+	TEE_Free(attr);
+	if (obj->attribs_hdl == TEE_HANDLE_NULL && hdl != TEE_HANDLE_NULL)
+		TEE_CloseObject(hdl);
```
`TEE_CloseObject()` should be able to handle `TEE_HANDLE_NULL` so I suppose there is no need to have `hdl != TEE_HANDLE_NULL`. Feel free to keep it if it is easier to understand it that way.

`obj->attrib_hdl == TEE_HANDLE_NULL` seems to be here for "re-entry" check that `hdl` does not get freed while in use.

etienne-lms replied:
agree, `if (obj->attribs_hdl == TEE_HANDLE_NULL)` is enough.

ruchi393 replied:
Agree, will change in the review patchset

vesajaaskelainen commented on ta/pkcs11/src/object.c:
```diff
@@ -427,3 +427,552 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+
+	for (; cur < end; cur += len) {
+		struct pkcs11_attribute_head *cli_ref =
+			(struct pkcs11_attribute_head *)(void *)cur;
```
double cast in here -- one should be sufficient?

etienne-lms replied:
agree

etienne-lms commented on ta/pkcs11/src/object.c:
```diff
@@ -427,3 +427,552 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+	 * session objects (that are visible to the session). Then scan all
+	 * remaining persistent object for which no session object handle was
+	 * published to the client.
+	 */
```
I think we can remove the 2nd half:
```c
	/*
	 * Scan all objects (sessions and persistent ones) and set a list of
	 * candidates that match caller attributes.
	 */
```

ruchi393 replied:
Changed

etienne-lms commented on ta/pkcs11/src/object.c:
```diff
@@ -427,3 +427,552 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+		/*
+		 * If there are no attributes specified, we return
+		 * every object
+		 */
```
we can remove this inline comment.

ruchi393 replied:
Removed

etienne-lms commented on ta/pkcs11/src/object.c:
```diff
@@ -427,3 +427,552 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+	if (rc == PKCS11_RV_NOT_FOUND)
+		rc = PKCS11_CKR_OK;
+
+	/* Save target attributes to search (if needed later) */
```
maybe remove this inline comment. I don't find it really useful.

ruchi393 replied:
Removed

etienne-lms commented on ta/pkcs11/src/object.c:
```diff
@@ -427,3 +427,552 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+	}
+
+	if (rc == PKCS11_RV_NOT_FOUND)
+		rc = PKCS11_CKR_OK;
```
We should remove these 2 lines.

ruchi393 replied:
Agree as we assign rc to PKCS11_CKR_OK after few lines in the same function so this seems redundant.

etienne-lms commented on ta/pkcs11/src/object.c:
```diff
@@ -427,3 +427,552 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+		goto out;
+	}
+
+	/* iterate over attributes and set their values */
```
s/iterate/Iterate/

ruchi393 replied:
Changed

etienne-lms commented:
looks all good to me, i'm ok for a squash-up.

ruchi393 commented:
> looks all good to me, i'm ok for a squash-up.

Thanks @etienne-lms . 
@vesajaaskelainen , can you have a look.

vesajaaskelainen commented:
> > looks all good to me, i'm ok for a squash-up.
> 
> Thanks @etienne-lms .
> @vesajaaskelainen , can you have a look.

Rebased it below my changes and tested it out.

Seems to be good.

Please do squashs.

ruchi393 commented:
@etienne-lms , for fixup commits which I have cherry-picked from your branch, do you want me to squash with relevant commits or leave them as it is
etienne-lms@642c7e4
etienne-lms@4509c65
etienne-lms@b5d5beb

etienne-lms commented:
better squash them into relevant commits.

etienne-lms commented on ta/pkcs11/src/pkcs11_token.h:
```diff
@@ -153,8 +153,8 @@ struct active_processing {
- * @handles - array of handle of matching objects (published handles)
- * @next - index of the next object handle to return to FindObject
+ * @handles - array of handle of matching objects
+ * @next - index of the next object handle to return to C_FindObject
```
very minor: this change in commit "ta: pkcs11: Add support for getting object size and attribute value" should move into the previous commit "ta: pkcs11: Add support for finding objects".

etienne-lms reviewed (commented):
LGTM but few fixup changes squashed in wrong commit

ruchi393 commented:
Fixed the issues with squashing. Thanks @etienne-lms for catching them.

ruchi393 commented:
@vesajaaskelainen , can you review and provide tags if ok.

vesajaaskelainen commented:
Re-tested them with my commits on top of them.

`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

Once merged I can start making PRs.

ruchi393 commented:
Tags applied. Thanks @vesajaaskelainen and @etienne-lms
